reframe/core/schedulers/sge.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2016-2021 Swiss National Supercomputing Centre (CSCS/ETH Zurich)
+# ReFrame Project Developers. See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+#
+# SGE backend
+#
+# - Initial version submitted by Mosè Giordano, UCL (based on the PBS backend)
+#
+
+import functools
+import re
+import time
+import xml.etree.ElementTree as ET
+
+import reframe.core.runtime as rt
+import reframe.utility.osext as osext
+from reframe.core.backends import register_scheduler
+from reframe.core.exceptions import JobSchedulerError
+from reframe.core.schedulers.pbs import PbsJobScheduler
+from reframe.utility import seconds_to_hms
+
+_run_strict = functools.partial(osext.run_command, check=True)
+
+
+@register_scheduler('sge')
+class SgeJobScheduler(PbsJobScheduler):
+    TASKS_OPT = ('-l select={num_nodes}:mpiprocs={num_tasks_per_node}'
+                 ':ncpus={num_cpus_per_node}')
+
+    def __init__(self):
+        self._prefix = '#$'
+        self._submit_timeout = rt.runtime().get_option(
+            f'schedulers/@{self.registered_name}/job_submit_timeout'
+        )
+
+    def _emit_lselect_option(self, job):
+        num_tasks_per_node = job.num_tasks_per_node or 1
+        num_cpus_per_task = job.num_cpus_per_task or 1
+        num_nodes = job.num_tasks // num_tasks_per_node
+        num_cpus_per_node = num_tasks_per_node * num_cpus_per_task
+        select_opt = ''
```

Comment:
You are ignoring completely the `num_nodes` etc. here. I don't know how practical this is? This backend can then only run single node jobs?

Reply:
I know :slightly_smiling_face:  This is the very last bit.  SGE does support multi-node jobs, the problem is more that, as far as I could understand, there isn't a general way to do it, the details depends on the setup of the system.

Reply:
Ok, I'm leaving it to @victorusu to have a further look, since he has some experience with SGE.

Reply:
@giordano, AFAIK SGE comes with the default parallel environments (`smp`, `make` and `mpi`) and they can be accessed using the `-pe` option. An example SGE preamble would be:
```
#$ -S /bin/bash
#$ -N example_job
#$ -o sge.out
#$ -e sge.err
#$ -pe smp 12
```
You are right that one can modify as much as you want and even remove those parallel environments. That exactly what I used to do back in the day... I used to have the `local`, `mpich`, `openmpi`, `mvapich` and `intelmpi` parallel environments set.

Depending on how you configure them, they can allocate multiple nodes, that what I used to do with the MPI-related parallel environments.

IMHO, the issue is that SGE doesn't know the concept of `tasks` and `cpus-per-tasks`. It only knows the concept of SLOTS. The number of SLOTS is passed after the parallel environment definition (in the example it is set to 12). 

Because of this limitation, people started to use custom launchers with SGE. A very famous one is called `ibrun` that was adopted by many US sites (https://github.com/TACC/lariat/blob/master/bin/ibrun.sge).

Fun fact: Back in the day, I wrote my own launcher that I named `srun` and it had the flags `-n` and `-c`. But at that time, I had no clue about the SLURM scheduler. 😀 That was a fun coincidence!

My proposal is to set the SLOTS to be configurable in the virtual partition.
We could have a virtual partition defined as:
```
{
    'name': 'smp',
    'time_limit': '10m',
    'scheduler': 'sge',
    'access': [
        f'-pe smp %(num_cpus_per_task)s'
    ],
    'name': 'mpi',
    'time_limit': '10m',
    'scheduler': 'sge',
    'access': [
        f'-pe mpi %(num_task)s'
    ],
    'name': 'mpismp',
    'time_limit': '10m',
    'scheduler': 'sge',
    'access': [
        f'-pe mpi %(num_task)s*%(num_cpus_per_task)s'
    ],
}
```

Please note that we need to have a syntax that is different than the one that I propose here because the product (`*` for the `mpismp` partition) must be evaluated. We cannot emit the preamble with the product.

Since the SGE scheduler doesn't have its own launcher, I think that It is the responsibility of the site to configure the launcher the way they want.
We could later introduce support for the `ibrun` launcher, for instance. But this is a different discussion because we are not sure which version of `ibrun` to support...

If that's the case, then virtual partition configuration could be
```
'scheduler': 'sge+ibrun',
```

What do you guys think?
How do you configure the parallel environments in your site?

Reply:
>  How do you configure the parallel environments in your site?

Yes, we do use parallel environments and the default one is called `mpi`, but as far as I could see, and you confirmed it, sysadmins can change the name of the environment or add new ones, which makes it harder to write a general script which can run on different systems.

Reply:
Thanks @victorusu for the insight and I didn't know that `ibrun` was to workaround SGE 😂 As a matter of fact, we do have an `ibrun` launcher inn ReFrame (I made it for TACC at some point). I am not sure though that the syntax you propose for the `access` is the correct. I was thinking that this configuration should be part of scheduler-specific configuration. `access` has the meaning of how you get access to that system partition, and here you overload that with something completely different I think. I would rather see it as follows:

```python
    'schedulers': [
        {
            'name': 'sge',
            'slots': [
                {
                    'name': 'smp',
                    'value': 'num_cpus_per_task'
                },
                {
                    'name': 'mpi',
                    'value': 'num_tasks'
                },
                {
                    'name': 'mpismp',
                    'value': 'num_tasks*num_cpus_per_task'
                }
            ],
            'target_systems': ['system_with_sge_scheduler']
        }
    ],
```

The SGE backend will then access the slots through the configuration:

```python
from reframe.core.runtime import runtime

slots = runtime.get_option('schedulers/@sge/slots')
```

From this point on, you can use this information to emit `-pe` command. What I am missing is what thing determines the slot to use. Any ideas?

To evaluate the expression for each slot, you can do this `eval(slot_value, job.__dict__)`.

NOTE: @victorusu I think that in your example the `-pe` option for the `mpismp` slot is wrong. It should be `-pe mpismp`, right?

Reply:
@giordano I think the slots concept that @victorusu described can be support from what ReFrame already offers with the concept of scheduler resources. Try define something along the following lines in your SGE scheduler partition (here called `foo`):

```python
                {
                    'name': 'foo',
                    'time_limit': '10m',
                    'scheduler': 'sge',
                    'resources': [
                        {
                            'name': 'smp',
                            'options': ['-pe', 'smp', '{num_slots}']
                        },
                        {
                            'name': 'mpi',
                            'options': ['-pe', 'mpi', '{num_slots}']
                        },
                        {
                            'name': 'mpismp',
                            'options': ['-pe', 'mpismp', '{num_slots}']
                        }
                    ]
                }
```

Each resource here is *slot* which then you can use in your test as follows:

```python
self.extra_resouces = {
    'smp': {'num_slots': self.num_cpus_per_task},
    'mpi': {'num_slots': self.num_tasks},
    'mpismp': {'num_slots': self.num_tasks*self.num_cpus_per_task}
}
```

@giordano Can you give it a try? If it works, then we only need to document this for the SGE scheduler and we can merge this PR.

Of course, you need to specify only the resources that you need. The important thing is that the test remains portable to systems with other schedulers, because if a partition does not define resources, `extra_resources` will be ignored.

Reply:
@vkarak, this is actually perfect, because this is a simple solution.

Reply:
Sorry for the delay, we had some issues on our clusters in the previous days, today they seem to be behaving better.

So, I think the problem is that at the moment this PR is not writing these options correctly, I get this in the preamble:
```sh
#$ :smp:1:mpi:4:mpismp:4
#$ -pe
#$ -pe
#$ -pe
```
which is clearly wrong.  As far as I understand, the first line comes from
https://github.com/eth-cscs/reframe/blob/d90431cacf792e7a586246ea0d5810a6fb93586d/reframe/core/schedulers/sge.py#L38-L63  I believe the function `_emit_lselect_option` doesn't make sense for SGE.  I'm missing what is writing the `-pe` lines (and the arguments don't seem to be passed correctly).  Any hints? :slightly_smiling_face:

Reply:
Yes, `_emit_lselect_option()` must be updated for the SGE scheduler. Since I have already started writing some unit tests for this, I can update this as well.

Reply:
Ciao @giordano, can you try again now? I removed the `-lselect` option (it has no meaning) and I'm supposed to emit correctly the options passed from the cmd line or the extra resources.

Reply:
Just to see how the job file is written, I'm using this test:
```py
import reframe as rfm
import reframe.utility.sanity as sn


@rfm.simple_test
class HelloTest(rfm.RegressionTest):
    valid_systems = ['*']
    valid_prog_environs = ['*']
    sourcepath = 'hello.c'
    executable_opts = ['> hello.out']
    sanity_patterns = sn.assert_found(r'Hello, World\!', 'hello.out')
    num_cpus_per_task = 1
    num_tasks = 4
    extra_resources = {
        'smp': {'num_slots': num_cpus_per_task},
        'mpi': {'num_slots': num_tasks},
        'mpismp': {'num_slots': num_tasks*num_cpus_per_task}
    }
```
but this is writing
```sh
#$ -pe
#$ smp
#$ 1
#$ -pe
#$ mpi
#$ 4
#$ -pe
#$ mpismp
#$ 4
```
which is wrong because it should be
```sh
#$ -pe smp 1
#$ -pe mi 4
#$ -pe mpismp 4
```

Reply:
My bad in the implementation 🤦 I will fix that now!